### ISSUE TITLE
The command will move slightly

### ISSUE THREAD

**Issue by Juddd** [powerline/powerline#1796]
I make a gif for specfiy this problem like following

![](https://i.stack.imgur.com/8wrdw.gif)

I don't sure this is a bug,but I hope this case get a improvement.Thanks

BTW,I'm in *Windows 10 Subsystem for Linux*

**Comment by ZyX-I** [comment]
The prompt is supposed to be colored. The WSL does not have 256 colors support?

**Comment by ZyX-I** [comment]
Also check the setting regarding ambiguous characters’ width: if terminal has one it should specify that ambiguous width characters have width equal to 1 (or single) (the only alternative should be 2 (or double)).

**Comment by Juddd** [comment]
As my search in web,somebody's prompt is colored indeed.But I don't sure it isnot in my case.I don't know how to adjust.

**Comment by ZyX-I** [comment]
I am not using WSL in general and don’t have access to Windows 10 now. The problem looks like one of the possible glitches due to something like ambiwidth terminal setting. And in any case if there are no 256 colors support then terminal emulator is not supported.

BTW, if this is the problem with ambiwidth then it should be “fixed” by running shell with

    POWERLINE_CONFIG_OVERRIDES=ext.shell.top_theme=ascii

environment variable (setting it in the working shell may work or may require additional actions to work depending on the setup, so it is better to just launch a new shell for test). Given that there is no coloring (do you have term_truecolor set by any chance? Sometimes it disables coloring just as effectively because it tries to use rarer supported 24-bit color escape sequences) the output should be pretty much the same, except that there will be no triangles.

**Comment by ZyX-I** [comment]
It looks like not only 256 colors, but 24-bit colors were added in later 2016 https://blogs.msdn.microsoft.com/commandline/2016/09/22/24-bit-color-in-the-windows-console/. Though given that Microsoft/BashOnWindows#345 was closed with a message which points to this article most likely there was no 256 colors support before that date.

**Comment by Juddd** [comment]
As I try,it is work on 24-bit
![](http://i.stack.imgur.com/4RRjk.png)

**Comment by ZyX-I** [comment]
Most likely this means that long sequences are not supported (powerline color sequences look like `^[[0;38;5;231;48;5;31;1m`: in *one* escape sequence do reset colors and attributes (0), set foreground (38;5;231), set background (48;5;31) and set some (AFAIR “bold”) attribute (1)).

And still what about a) ambiwidth terminal emulator setting: does it have it, what is it set to, etc and b) whether bug from your first post can be viewed when `top_theme` is `ascii`?

**Comment by Juddd** [comment]
Could I invent you into [this room](https://chat.stackoverflow.com/rooms/148183/for-solve-powerline-problem)?

**Comment by ZyX-I** [comment]
If long sequences are the case then

```Patch
diff --git a/powerline/renderers/shell/__init__.py b/powerline/renderers/shell/__init__.py
index ebb0501..93e15d3 100644
--- a/powerline/renderers/shell/__init__.py
+++ b/powerline/renderers/shell/__init__.py
@@ -112,52 +112,51 @@ class ShellRenderer(PromptRenderer):
 		False, the argument is reset to the terminal defaults. If an argument
 		is a valid color or attribute, it’s added to the ANSI escape code.
 		'''
-		ansi = [0]
+		ansis = [(0,)]
 		is_fbterm = self.used_term_escape_style == 'fbterm'
 		term_truecolor = not is_fbterm and self.term_truecolor
 		if fg is not None:
 			if fg is False or fg[0] is False:
-				ansi += [39]
+				ansis.append((39,))
 			else:
 				if term_truecolor:
-					ansi += [38, 2] + list(int_to_rgb(fg[1]))
+					ansis.append((38, 2) + int_to_rgb(fg[1]))
 				else:
-					ansi += [38, 5, fg[0]]
+					ansis.append((38, 5, fg[0]))
 		if bg is not None:
 			if bg is False or bg[0] is False:
-				ansi += [49]
+				ansis.append((49,))
 			else:
 				if term_truecolor:
-					ansi += [48, 2] + list(int_to_rgb(bg[1]))
+					ansis.append((48, 2) + int_to_rgb(bg[1]))
 				else:
-					ansi += [48, 5, bg[0]]
+					ansis.append((48, 5, bg[0]))
 		if attrs is not None:
 			if attrs is False:
-				ansi += [22]
+				ansis.append((22,))
 			else:
 				if attrs & ATTR_BOLD:
-					ansi += [1]
+					ansis.append((1,))
 				elif attrs & ATTR_ITALIC:
 					# Note: is likely not to work or even be inverse in place of
 					# italic. Omit using this in colorschemes.
-					ansi += [3]
+					ansis.append((3,))
 				elif attrs & ATTR_UNDERLINE:
-					ansi += [4]
+					ansis.append((4,))
 		if is_fbterm:
 			r = []
-			while ansi:
-				cur_ansi = ansi.pop(0)
-				if cur_ansi == 38:
-					ansi.pop(0)
-					r.append('\033[1;{0}}}'.format(ansi.pop(0)))
-				elif cur_ansi == 48:
-					ansi.pop(0)
-					r.append('\033[2;{0}}}'.format(ansi.pop(0)))
+			for ansi in ansis:
+				if ansi[0] == 38:
+					r.append('\033[1;{0}}}'.format(ansi[1]))
+				elif ansi[0] == 48:
+					r.append('\033[2;{0}}}'.format(ansi[1]))
 				else:
-					r.append('\033[{0}m'.format(cur_ansi))
+					r.append('\033[{0}m'.format(ansi))
 			r = ''.join(r)
 		else:
-			r = '\033[{0}m'.format(';'.join(str(attr) for attr in ansi))
+			r = ''.join((
+				'\033[{0}m'.format(';'.join(str(attr) for attr in ansi))
+				for ansi in ansis))
 		if self.tmux_escape:
 			r = '\033Ptmux;' + r.replace('\033', '\033\033') + '\033\\'
 		elif self.screen_escape:
```

should solve the problem.

**Comment by ZyX-I** [comment]
No, I do not like the idea of using a chat, especially given that I tend to write long messages.

**Comment by ZyX-I** [comment]
And issue tracker is a thing which allows me to go do something else in the middle of the conversation without notice.

**Comment by Juddd** [comment]
I'm in latest version of windows.I doubt it don't support long sequences.If it is dont' support indeed.Where should place [your code here](https://github.com/powerline/powerline/issues/1796#issuecomment-312516175)?

**Comment by ZyX-I** [comment]
@Juddd It is a patch, you supply it to the `patch` command (`patch -p1` while current working directory is repository root). Diff headers (since in that particular patch there is only one, on the first line of the code block) tell it which file it needs to edit.

**Comment by ZyX-I** [comment]
@Juddd And latest means nothing. There was a time when latest version of Windows *did not support 256 colors*. And it was very easy to predict this request for anybody who invested a little time into researching how linux users customize their shells.

**Comment by ZyX-I** [comment]
Though note that my `ag` version is 1.0.2 (Gentoo amd64 (stable)) while latest version in portage is 2.0.0. Maybe something was fixed since then.
